### ISSUE TITLE
Consolidate the Setup/SetupSequence/Verify "overridable member" duplication into a MoqVerificationHelpers helper; align Moq1203's private symbol extraction

### ISSUE THREAD

**Issue by rjmurillo** [rjmurillo/moq.analyzers#1274]
## 1. Summary

An internal code audit found that the three "should be used only for overridable members" analyzers (Moq1200 Setup, Moq1207 SetupSequence, Moq1210 Verify) duplicate the same core: extract the mocked member symbol from the setup/verify lambda, bail on interface members, bail on overridable/allowed members, then report. Additionally, `MethodSetupShouldSpecifyReturnValueAnalyzer` (Moq1203) privately re-implements `MoqVerificationHelpers.TryGetMockedMemberSymbol` minus its `IDelegateCreationOperation` unwrap. This issue extracts one shared helper and (as a separately-flagged step) switches Moq1203 to the Common helper.

## 2. Current state (evidence)

**The duplicated trio core:**

- `src/Analyzers/SetupShouldBeUsedOnlyForOverridableMembersAnalyzer.cs:75-100` — `IsSetupOnNonOverridableMember`: gate (`IsMoqSetupMethod || IsMoqEventSetupMethod`), then

```csharp
ISymbol? candidate = MoqVerificationHelpers.TryGetMockedMemberSymbol(invocationOperation);
if (candidate is null
    || candidate.ContainingType?.TypeKind == TypeKind.Interface
    || candidate.IsOverridableOrAllowedMockMember(knownSymbols))
{
    return false;
}
```

- `src/Analyzers/SetupSequenceShouldBeUsedOnlyForOverridableMembersAnalyzer.cs:50-76` — `AnalyzeInvocation`: gate (`IsMoqSetupSequenceMethod`), then the identical null/interface/overridable triple bail at lines 62–68.
- `src/Analyzers/VerifyShouldBeUsedOnlyForOverridableMembersAnalyzer.cs:50-116` — `AnalyzeInvocation` + `IsInterfaceMember` + `IsMemberAllowedForVerification`: gate (`IsMoqVerificationMethod` minus `VerifyNoOtherCalls`, lines 90–100), the same extraction via `MoqVerificationHelpers.TryGetMockedMemberSymbol` for the non-`VerifySet` path (line 74), the same interface bail (lines 102–105), and the same `IsOverridableOrAllowedMockMember` check for the non-`VerifySet` path (line 115).

Roughly 50 of the trio's ~90 body lines are the same logic written three ways.

**The Moq1203 re-implementation:**

- `src/Analyzers/MethodSetupShouldSpecifyReturnValueAnalyzer.cs:121-142` — `TryGetMockedMethodSymbol` + `TryGetSetupArgument`:

```csharp
private static ISymbol? TryGetMockedMethodSymbol(IInvocationOperation moqSetupInvocation) =>
    TryGetSetupArgument(moqSetupInvocation)?.GetReferencedMemberSymbolFromLambda();

private static IOperation? TryGetSetupArgument(IInvocationOperation moqSetupInvocation)
{
    if (moqSetupInvocation.Arguments.Length == 0)
    {
        return null;
    }

    IOperation argumentOperation = moqSetupInvocation.Arguments[0].Value;

    // Unwrap conversions (Roslyn often wraps lambdas in IConversionOperation)
    argumentOperation = argumentOperation.WalkDownImplicitConversion();

    return argumentOperation is IAnonymousFunctionOperation lambdaOperation
        ? lambdaOperation.Body
        : null;
}
```

This is `MoqVerificationHelpers.TryGetMockedMemberSymbol` (`src/Common/MoqVerificationHelpers.cs:58-70`) minus the `IDelegateCreationOperation` unwrap that `ExtractLambdaFromArgument` (`MoqVerificationHelpers.cs:15-27`) performs.

## 3. Desired state

- One helper in `src/Common/MoqVerificationHelpers.cs` encapsulates "extract mocked member + interface bail + overridability bail"; the trio calls it. Each analyzer keeps its own invocation gate, diagnostic location, and message formatting (see Behavioral notes — these differ deliberately and are pinned by tests).
- Moq1203 uses `MoqVerificationHelpers.TryGetMockedMemberSymbol` instead of its private copy (flagged step, see section 5a).

## 4. Implementation plan

1. **Add the shared helper** to `src/Common/MoqVerificationHelpers.cs` (add `using System.Diagnostics.CodeAnalysis;` at the top; the file currently only imports `Microsoft.CodeAnalysis.Operations`):

```csharp
    /// <summary>
    /// Extracts the mocked member from a Moq Setup/SetupSequence/Verify invocation and
    /// determines whether it is a member that Moq cannot mock (not overridable).
    /// </summary>
    /// <param name="moqInvocation">The Setup/SetupSequence/Verify invocation operation.</param>
    /// <param name="knownSymbols">The known Moq symbols for this compilation.</param>
    /// <param name="mockedMemberSymbol">
    /// When this method returns <see langword="true"/>, the non-overridable mocked member.
    /// </param>
    /// <returns>
    /// <see langword="true"/> when the invocation's lambda references a member that is
    /// neither an interface member nor overridable/allowed for mocking; otherwise
    /// <see langword="false"/> (including when no member could be extracted).
    /// </returns>
    public static bool TryGetNonOverridableMockedMember(
        IInvocationOperation moqInvocation,
        MoqKnownSymbols knownSymbols,
        [NotNullWhen(true)] out ISymbol? mockedMemberSymbol)
    {
        mockedMemberSymbol = null;

        ISymbol? candidate = TryGetMockedMemberSymbol(moqInvocation);
        if (candidate is null
            || candidate.ContainingType?.TypeKind == TypeKind.Interface
            || candidate.IsOverridableOrAllowedMockMember(knownSymbols))
        {
            return false;
        }

        mockedMemberSymbol = candidate;
        return true;
    }
```

2. **Migrate `SetupShouldBeUsedOnlyForOverridableMembersAnalyzer`** (representative before/after). Delete `IsSetupOnNonOverridableMember` (lines 68–100) and inline the gate into `AnalyzeInvocation`:

   Before (`AnalyzeInvocation`, lines 52–66 + helper 75–100):

```csharp
    private static void AnalyzeInvocation(OperationAnalysisContext context, MoqKnownSymbols knownSymbols)
    {
        if (context.Operation is not IInvocationOperation invocationOperation)
        {
            return;
        }

        if (!IsSetupOnNonOverridableMember(invocationOperation, knownSymbols, out ISymbol? mockedMemberSymbol))
        {
            return;
        }

        Diagnostic diagnostic = invocationOperation.Syntax.CreateDiagnostic(Rule, mockedMemberSymbol.ToDisplayString());
        context.ReportDiagnostic(diagnostic);
    }
```

   After:

```csharp
    private static void AnalyzeInvocation(OperationAnalysisContext context, MoqKnownSymbols knownSymbols)
    {
        if (context.Operation is not IInvocationOperation invocationOperation)
        {
            return;
        }

        IMethodSymbol targetMethod = invocationOperation.TargetMethod;

        // Check if the invoked method is a Moq Setup method.
        if (!targetMethod.IsMoqSetupMethod(knownSymbols) && !targetMethod.IsMoqEventSetupMethod(knownSymbols))
        {
            return;
        }

        if (!MoqVerificationHelpers.TryGetNonOverridableMockedMember(invocationOperation, knownSymbols, out ISymbol? mockedMemberSymbol))
        {
            return;
        }

        // Moq1200 reports at the whole invocation with the member's display string (pinned by tests).
        Diagnostic diagnostic = invocationOperation.Syntax.CreateDiagnostic(Rule, mockedMemberSymbol.ToDisplayString());
        context.ReportDiagnostic(diagnostic);
    }
```

   Repeat for:
   - `src/Analyzers/SetupSequenceShouldBeUsedOnlyForOverridableMembersAnalyzer.cs` — replace lines 62–68 with the helper call; keep the `IsMoqSetupSequenceMethod` gate, and keep the member-syntax location logic at lines 70–74 exactly as-is.
   - `src/Analyzers/VerifyShouldBeUsedOnlyForOverridableMembersAnalyzer.cs` — only the **non-`VerifySet`** path changes: replace `mockedMemberSymbol = MoqVerificationHelpers.TryGetMockedMemberSymbol(invocationOperation);` (line 74) plus the follow-up `IsInterfaceMember`/`IsMemberAllowedForVerification` checks (lines 77–85, non-VerifySet arm at line 115) with a single `TryGetNonOverridableMockedMember` call. The `VerifySet` path (lines 65–71: `ExtractLambdaFromArgument` + `ExtractPropertyFromVerifySetLambda`, and the `IPropertySymbol.IsOverridable()` allowed-check at lines 110–113) must stay verbatim — `VerifySet` semantics (setter overridability) intentionally differ. After the refactor `IsInterfaceMember` and `IsMemberAllowedForVerification` remain only for the VerifySet arm; simplify them accordingly or keep them intact — do not change what executes.

3. **Flagged step — Moq1203 alignment (E-6).** In `src/Analyzers/MethodSetupShouldSpecifyReturnValueAnalyzer.cs`, replace the body of `TryGetMockedMethodSymbol` (lines 121–122) with:

```csharp
    private static ISymbol? TryGetMockedMethodSymbol(IInvocationOperation moqSetupInvocation) =>
        MoqVerificationHelpers.TryGetMockedMemberSymbol(moqSetupInvocation);
```

   and delete `TryGetSetupArgument` (lines 124–142). See section 5a before doing this.

4. Run `dotnet format`, `dotnet build`, full test suite.

## 5. Behavioral notes

**Deliberate differences that MUST be preserved** (spans and messages are pinned by markup tests; do NOT normalize them):

| Rule | Diagnostic span | Message argument |
|---|---|---|
| Moq1200 (Setup) | whole invocation — tests pin `{|Moq1200:new Mock<SampleClass>().Setup(x => x.Property)|}` (`tests/Moq.Analyzers.Test/SetupShouldBeUsedOnlyForOverridableMembersAnalyzerTests.cs:12-25`) | `mockedMemberSymbol.ToDisplayString()` |
| Moq1207 (SetupSequence) | the member syntax inside the lambda — tests pin `SetupSequence(x => {|Moq1207:x.Property|})` (`tests/Moq.Analyzers.Test/SetupSequenceShouldBeUsedOnlyForOverridableMembersAnalyzerTests.cs:25-29`), falling back to the whole invocation when the member syntax is unavailable (`SetupSequenceShouldBeUsedOnlyForOverridableMembersAnalyzer.cs:70-74`) | `mockedMemberSymbol.ToDisplayString()` |
| Moq1210 (Verify) | whole invocation — tests pin `{|Moq1210:new Mock<SampleClass>().Verify(x => x.Property)|}` (`tests/Moq.Analyzers.Test/VerifyShouldBeUsedOnlyForOverridableMembersAnalyzerTests.cs:13-30`) | `mockedMemberSymbol.Name` (`VerifyShouldBeUsedOnlyForOverridableMembersAnalyzer.cs:120`) |

Consequences of normalizing (why we do not): moving Moq1200/Moq1210 to the member syntax, or Moq1207 to the whole invocation, changes every pinned span in three test suites; changing Moq1210's `Name` to `ToDisplayString()` changes the rendered message from e.g. `'Calculate'` to `'SampleClass.Calculate(int, int, int)'`. Both would be user-visible churn (baselines, suppression review) with zero correctness gain. The helper therefore returns the symbol and leaves location/format to each analyzer.

**5a. Flagged: Moq1203 switch is a potential behavior change — treat as its own commit with its own test evidence.**

- Delta: `MoqVerificationHelpers.TryGetMockedMemberSymbol` additionally unwraps `IDelegateCreationOperation` (`MoqVerificationHelpers.cs:20-23`); the private Moq1203 copy does not. The Common helper is a strict superset: wherever the private copy finds a lambda, the helper finds the same lambda; the only new inputs are arguments where Roslyn wraps the lambda in a delegate creation.
- Why this is expected to be a no-op for Moq1203: the analyzer only proceeds for `IsMoqSetupMethod` targets (`ISymbolExtensions.Moq.cs:40-43` — `Mock<T>.Setup<TResult>(Expression<Func<T,TResult>>)`, generic method only). Expression-tree-typed lambda arguments surface as `IConversionOperation` over `IAnonymousFunctionOperation` (already handled by both versions via `WalkDownImplicitConversion`); `IDelegateCreationOperation` arises for delegate-typed parameters (e.g. Moq's `SetupSet(Action<T>)` — a method Moq1203 never matches). The near-exhaustive Moq1203 suite (`tests/Moq.Analyzers.Test/MethodSetupShouldSpecifyReturnValueAnalyzerTests.cs`, 14 fact/theory members expanded across the namespace/Moq-version matrix) passing unchanged is the proof.
- If it is not a no-op, the direction of change is: Moq1203 starts *detecting* setups it previously silently skipped (a fix, not a regression), because a `null` member symbol means "no diagnostic".
- Test plan for this step:
  1. Run only the Moq1203 suite before and after; require identical pass/fail and identical diagnostic expectations.
  2. Add one regression row per shape to the Moq1203 test data if not already present: block-bodied lambda setup, method-group chained `Returns`, and a `SetupSet`-style delegate lambda on a returning member (expected: still no Moq1203, since `SetupSet` is not a Setup method).
  3. If any existing test changes outcome, revert this step, keep the private copy, and record the differing input shape in the issue before proceeding.

## 6. Constraints

- `netstandard2.0`, Roslyn pinned 4.8, no new dependencies.
- Behavior-preserving except for the explicitly flagged 5a step; the full existing test suite is the regression gate.
- The trio analyzers keep `[DiagnosticAnalyzer(LanguageNames.CSharp)]`; no abstract analyzer class is introduced by this issue (the helper is a static method precisely to avoid base-class/attribute concerns).
- `MoqVerificationHelpers` lives in the shared `Common.projitems` compiled into both `Moq.Analyzers` and `Moq.CodeFixes`; keep the helper `public static` inside the `internal static class` like its siblings.

## 7. Validation

1. `dotnet format` — clean.
2. `dotnet build` — 0 warnings, 0 errors.
3. `dotnet test --settings ./build/targets/tests/test.runsettings` — full suite green; specifically zero changes in `SetupShouldBeUsedOnlyForOverridableMembersAnalyzerTests`, `SetupSequenceShouldBeUsedOnlyForOverridableMembersAnalyzerTests`, `VerifyShouldBeUsedOnlyForOverridableMembersAnalyzerTests`, `MethodSetupShouldSpecifyReturnValueAnalyzerTests`, and `VerifyOverridableMembersFixer`-related code-fix tests (the fixer at `src/CodeFixes/VerifyOverridableMembersFixer.cs:57` also consumes `TryGetMockedMemberSymbol` — untouched here, but exercised by the suite).

## 8. Acceptance criteria

- [ ] `MoqVerificationHelpers.TryGetNonOverridableMockedMember` exists with `[NotNullWhen(true)]` out parameter and is used by all three trio analyzers.
- [ ] The null/interface/overridable triple-bail appears exactly once in `src/` (in the helper).
- [ ] Moq1200/Moq1207/Moq1210 spans and message arguments are byte-identical to before (test markup unchanged).
- [ ] Moq1203 uses the Common helper (or, if step 5a surfaced a real behavior delta, the issue documents the input shape and the step is deferred).
- [ ] `MethodSetupShouldSpecifyReturnValueAnalyzer.TryGetSetupArgument` is deleted (unless 5a deferred).
- [ ] Full test suite passes.

## 9. Out of scope + Related issues

Out of scope: the `Initialize` boilerplate consolidation (companion issue #1273); the argument-by-ordinal robustness gap in `MoqVerificationHelpers`/Verify (`Arguments[0]` assumes source order — separate audit finding); the Moq1200 static-member test fixture gap noted by the audit's test pass.

Related: #627 / #632 (CRAP reduction — three copies of the same branch logic each carry their own partially-covered branches); companion issue #1273 (same files, different layer — rebase whichever lands second; conflicts are trivial since this issue touches analyzer bodies and 27 touches `Initialize`).

**Comment by rjmurillo** [comment]
Resolved by #1311 (MERGED 2026-07-06). Verified on origin/main: `src/Common/MoqVerificationHelpers.cs` exists and consolidates the Setup/SetupSequence/Verify "overridable member" duplication; it is consumed by MethodSetupShouldSpecifyReturnValueAnalyzer, NoMethodsInPropertySetupAnalyzer, SetupSequenceShouldBeUsedOnlyForOverridableMembersAnalyzer, SetupShouldBeUsedOnlyForOverridableMembersAnalyzer, VerifyShouldBeUsedOnlyForOverridableMembersAnalyzer, and VerifyOverridableMembersFixer (Moq1203 private-symbol extraction aligned to the shared path). This issue stayed open only because #1311's body used `Fixes #1273, #1274, ...` — GitHub applies the auto-close keyword to the first issue in a comma list only. Closing manually.
